src/raftstore/store/peer.rs

```diff
@@ -498,25 +563,34 @@ impl Peer {
         };
 
         self.raft_group.apply_conf_change(conf_change);
+        self.repropose_conf_change();
+
         res
     }
 
-    fn repropose_pending_cmds(&mut self) -> Result<()> {
-        if !self.pending_cmds.is_empty() {
-            info!("re-propose {} pending commands after empty entry",
-                  self.pending_cmds.len());
-            // TODO: use a better way to avoid clone.
-            let mut cmds: Vec<PendingCmd> = Vec::with_capacity(self.pending_cmds.len());
-            for cmd in self.pending_cmds.values() {
-                // We only need cmd for later re-propose.
-                cmds.push(PendingCmd { cmd: cmd.cmd.clone(), ..Default::default() });
+    fn find_cb(&mut self, uuid: Uuid, term: u64, cmd: &RaftCmdRequest) -> Option<Callback> {
+        let is_conf_change = get_change_peer_cmd(cmd).is_some();
+        loop {
+            let cmd = if is_conf_change {
+                self.pending_cmds.pop_conf_change()
+            } else {
+                self.pending_cmds.pop_normal()
+            };
+            if cmd.is_none() {
+                return None;
             }
-
-            for mut cmd in &mut cmds {
-                try!(self.propose_pending_cmd(&mut cmd));
+            let head = cmd.unwrap();
+            if head.uuid == uuid {
+                return Some(head.cb);
+            }
+            assert!(!self.is_leader(), "leader should not have stale callback.");
```

Comment:
Then it will commit and apply from last time again, so there will be no stale callback and this line won't be reached.





On Thu, Apr 28, 2016 at 5:06 PM -0700, "siddontang" <notifications@github.com> wrote:












In src/raftstore/store/peer.rs:
>              }
> -
> -            for mut cmd in &mut cmds {
> -                try!(self.propose_pending_cmd(&mut cmd));
> +            let head = cmd.unwrap();
> +            if head.uuid == uuid {
> +                return Some(head.cb);
> +            }
> +            assert!(!self.is_leader(), "leader should not have stale callback.");



how about for 4, the origin leader become leader again?



—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub